### DISCUSSION TITLE
Common topic : UI Values for new record

### DISCUSSION THREAD

**Discussion by TonyGravagno** [colinhacks/zod#2431] [Q&A]
I've seen this asked a number of ways but I don't know if there is a common defined/accepted pattern. (I'm relatively new to TS and Zod, please be gentle 😄 )

Consider a form to enter a new record - a person, company, product. The default consists of a bunch of empty fields, undefined, zeros, etc. But those are almost certainly not valid `.safeParse()` values. The schema should not specify .optional() or .default(0), but that is indeed what we want to use as the default value in the UI. 

My approach to this has been to derive defaults based on the schema, and I've just started to use code that generates a default object from schema: https://github.com/colinhacks/zod/discussions/1953

The intent is to set a new data object to these values and then only show errors when the invalid default is left untouched by the user.

I just want to confirm : What is the best/right pattern to set values that are not schema-compliant? When using React Hook Forms, for example, I want to ensure that the lifecycle of field entry does not flag a field as being bad before a user has a chance to modify it. Is there a way to tell Zod "this is the default but this value should not be accepted in validation" ?

 Common examples:

- Required Name field is initially `undefined`
- Age field is initially set to undefined or zero but must be numeric `.gt(0)`
- Date field needs to be set to something but is initially `undefined`

Thanks for you patience. 😄 

EDIT: Ref also : https://github.com/colinhacks/zod/issues/1206

**Comment by scotttrinh** [comment]
I do not speak for the whole project, but I do have some personal thoughts related to this friction:

I think, in general, the issue as stated should be addressed within the context of the tool that is using Zod rather than in Zod itself. Every additional feature and workflow-specific flow that we abstract over means that Zod itself _grows_ in complexity. We've committed to a certain amount of complexity in providing tools like `coerce`, `transform`, `preprocess`, but in every case, Zod pays the cost in becoming slightly harder to maintain for the maintainers and slightly harder for people to understand when they're coming to Zod.

The underlying principle with Zod is to give you runtime representation of your compile-time types, and enhance those representations with useful invariants and transformations that exist at runtime. It's specifically _not_ a form validation library, or an API validation library, or a query string parser, or a JSON parser, or any of the myriad of use cases that we all use it for. So the fact that a form library (any form library as far as I can tell!) doesn't integrate with that runtime system well in some common but complex cases is definitely something I think should be improved at the form-library level. Anything we can do to make that integration easier, more consistent, and that helps other integrations is definitely something I think we should continue to take on, but Zod shouldn't know anything about how HTML forms function.

---

Having said that, some specific notes since most of these problems come down to a few key areas:
- _When_ the form library chooses to parse the input
- How the form library responds to `ZodEffects` (refine, transform, et al)
- What kind of processing the library does to the inputs (which are all and everywhere `string`s, not `undefined`, or `null`, or `0`)
- How the library populates the initial form control values

I do not myself maintain any form libraries, so I can't speak to the complexities here, but most of the questions I've seen usually hit a couple of these issues. I think that's why we haven't really hit on a catch-all solution to works for people struggling with Zod and different form libraries.

**Comment by TonyGravagno** [comment]
@scotttrinh - Thanks for your considered response. As a developer of over 45 years, I'm completely onboard with avoiding feature-creep and out-of-scope bolt-ons.

As a relative newcomer to Zod, reading as much as I can to get up to speed with the functionality and ecosystem, I'm on the same trajectory as anyone else, and it's similar to the path we follow for all FOSS : The documentation covers specific properties and functions pretty well, but so far I'm not understanding the higher level flow of Zod. I don't yet understand the .input and .output mechanisms, or exactly how or when to hook into the pre- and post- transformations. I'm not sure what other libraries do "for" us in this respect, or where they are not doing things for us, so I don't know where the lines are yet. And in short, I don't know what I don't know about what's available and where to plug in.

Learning these nuances requires an ongoing investment of time - we do what we need to do. As we've seen, this area that I'm asking about is a common area of confusion. It's not related to specific form libraries or APIs. As of now I'm looking for articles or videos on Zod internals, in-depth dive-in kind of stuff, short of reading all of the code. (Anyone have links? Anyone planning on doing this?)

The more I think about the topic at hand, the more I understand where the (my) confusion is. Quite simply: Don't parse the record until you know there's been an opportunity to modify all of the fields. This applies to forms or any back-end framework that builds a record from discrete modules. We can create a default data object from the schema, and it doesn't matter if that's valid. We know it's not. So don't parse it. After all of the processing is done, **then** parse it to see if it passes schema validation. As I'm understanding the Error message mapping, it's probably best to map error codes, not text, so that we can switch on codes, and then use translations to get the actual text for a UI if that's required.

Anyone want to chime in on that? ( Or are we at "what the heck did he just say?" )

Thanks again.
